File: lab3_Parsa/cat.py
import sys

def cat(n, b, s, args):
    open_files = []
    if len(args) == 0:
        open_files.append(sys.stdin)
    else:
        for arg in args:
            open_files.append(open(arg, 'r'))

    for file in open_files:
        line_cnt = 0
        empty_lines = 0 # holds the number of consecutive empty lines, used for s flag
        for line in file:
            # regardless of the flags(s, n, and b), when the line is not empty (i.e. != '\n')
            # we have to print it
            if line != '\n':
                empty_lines = 0 # reset number of consecutive empty lines
                # if n or b is specified we have to show the line number
                if n or b:
                    line_cnt += 1
                    sys.stdout.write("%6d  %s"%(line_cnt, line))
                else:
                    #otherwise just the line
                    sys.stdout.write(line)
            else:
                #increment the number of consecutive empty lines
                empty_lines += 1
                # if s flag is specified, we can only print the empty line, if the number of consecutive
                # empty lines is 1, if s is not specified, then we can print the empty line.
                if not s or empty_lines == 1:
                    if n:
                        line_cnt += 1
                        sys.stdout.write('%6d\n'%(line_cnt))
                    else:
                        sys.stdout.write(line)

    if len(args) != 0:
        for file in open_files:
            file.close()

def main():
    nflag = bflag = sflag = False
    cntr = 1

    for i in range(1, len(sys.argv)):
        if sys.argv[i][0] == '-':
            for j in range(1, len(sys.argv[i])):
                if sys.argv[i][j] == 'n':
                    nflag = True
                elif sys.argv[i][j] == 'b':
                    bflag = True
                elif sys.argv[i][j] == 's':
                    sflag = True
                else:
                    sys.stderr.write("unknown option -- {}\n".format(sys.argv[i][j]))

        else:
            break
        cntr += 1

    cat(nflag, bflag, sflag, sys.argv[cntr:])


if __name__ == '__main__':
    main()

File: lab3_Parsa/cat.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

/*
 *Synopsis:
 *
 *	cat [-nbs] [files ...]
 * */

void cat(FILE *fp, int n, int b, int s) {
	
}

int main(int argc, char *argv[]) {
	int n = 0, b = 0, s = 0;
	//./a.out -ns somfile
	int i = 1;
	for(; i < argc; i++){
		if(argv[i][0] == '-') {
			for (int j = 1; argv[i][j] != '\0'; j++) {
				if(argv[i][j] == 'n') {
					n = 1;
				} else if(argv[i][j] == 's') {
					s = 1;
				} else if(argv[i][j] == 'b') {
					b = 1;
				} else {
					fprintf(stderr, "%s: unknown option -- %c\n", argv[0], argv[i][j]);
					exit(1);
				}

			}
		} else {
			break;
		}
	}
	
	if (i == argc) {
		cat(stdin, n, b, s);
	}

	for(; i < argc; i++) {
		FILE *fp = fopen(argv[i], "r");
		if (fp == NULL) {
			fprintf(stderr, "%s: no such file or directory %s\n", argv[0], argv[i] );
			exit(1);
		}
		cat(fp, n, b, s);
		//I have to close the file now, because it occupies memory, and I have to release that memory
	}
}
